datasets/census_bureau_acs/_images/run_csv_transform_kub/csv_transform.py

```diff
@@ -0,0 +1,550 @@
+        "54": "West Virginia",
+        "55": "Wisconsin",
+        "56": "Wyoming",
+        "60": "American Samoa",
+        "66": "Guam",
+        "69": "Commonwealth of the Northern Mariana Islands",
+        "72": "Puerto Rico",
+        "78": "United States Virgin Islands",
+    }
+
+    logging.info("Extracting the data from API and loading into dataframe...")
+    if report_level == "national_level":
+        df = extract_data_and_convert_to_df_national_level(
+            group_id, year_report, api_naming_convention, source_url
+        )
+    elif report_level == "state_level":
+        df = extract_data_and_convert_to_df_state_level(
+            group_id, state_code, year_report, api_naming_convention, source_url
+        )
+
+    logging.info("Replacing values...")
+    df = df.replace(to_replace={"KPI_Name": group_id})
+
+    logging.info("Renaming headers...")
+    rename_headers(df, rename_mappings)
+
+    logging.info("Creating column geo_id...")
+    if geography == "censustract" or geography == "blockgroup":
+        df["tract"] = df["tract"].apply(change_length, args=("6"))
+        df["state"] = df["state"].apply(change_length, args=("2"))
+        df["county"] = df["county"].apply(change_length, args=("3"))
+
+    df = create_geo_id(df, concat_col)
+
+    logging.info("Pivoting the dataframe...")
+    df = df[["geo_id", "KPI_Name", "KPI_Value"]]
+    df = df.pivot_table(
+        index="geo_id", columns="KPI_Name", values="KPI_Value", aggfunc=np.sum
+    ).reset_index()
+
+    logging.info("Reordering headers...")
+    df = df[headers]
+
+    logging.info(f"Saving to output file.. {target_file}")
+    try:
+        save_to_new_file(df, file_path=str(target_file))
+    except Exception as e:
+        logging.error(f"Error saving output file: {e}.")
+
+    logging.info(
+        f"Uploading output file to.. gs://{target_gcs_bucket}/{target_gcs_path}"
+    )
+    upload_file_to_gcs(target_file, target_gcs_bucket, target_gcs_path)
+
+    logging.info(
+        f"ACS {pipeline_name} process completed at "
+        + str(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
+    )
+
+
+def string_replace(source_url, replace: dict) -> str:
+    for k, v in replace.items():
+        source_url_new = source_url.replace(k, v)
+    return source_url_new
+
+
+def extract_data_and_convert_to_df_national_level(
+    group_id: dict, year_report: str, api_naming_convention: str, source_url: str
+) -> pd.DataFrame:
+    list_temp = []
+    for key in group_id:
+        logging.info(f"reading data from API for KPI {key}...")
+        # replece = {
+        #     "+year_report+": year_report,
+        #     "+key[0:-3]+": key[0:-3],
+        #     "+key[-3:]+": key[-3:],
+        #     "+api_naming_convention+": api_naming_convention,
+        # }
+        # source_url_new = string_replace(source_url, replece)
+        str1 = source_url.replace("+year_report+", year_report)
+        str2 = str1.replace("+key[0:-3]+", key[0:-3])
```

Comment:
Use descriptive names for these character ranges

Reply:
done